# 1: Introduktion

Velkommen til forløbet om programmering!

Når du programmerer, så fortæller du en computer, hvad den skal gøre. Du
kommunikerer med computeren gennem kommandoer.

Kommandoerne er ligesom legoklodser. Der er mange forskellige, men de
kan ikke bruges til meget, når de står alene. Kombinerer du i stedet flere
legoklodser, så kan du bygge næsten alt, og kun fantasien sætter grænser.

Programmering fungerer på samme måde. Alt software i verden er bygget
af de samme fundamentale byggeklodser. Når du lærer at kombinere dem
på en smart måde, så kan du programmere næsten alt.

Python er et programmeringssprog. Computere kan ikke snakke hverken
dansk eller engelsk, og derfor skal vi bruge et sprog, som computeren forstår. På Machine Learning Camp vælger vi at bruge Python til at programmere i Jupyter Notebooks gennem Google Colab. Her i Jupyter Notebooks er kode og tekst delt op i såkaldte "celler", som kan køres ved at man markererer dem og derefter enten trykker på pilen i top baren eller bruger tasteturgenvejen CTRL + ENTER. I topbaren er der en drop-down menu, der for hver celle angiver om cellen er en "Markdown" celle eller en "Code" celle. Markdown celler er tekst, mens Code celler indeholder kode, der kan køres. Prøv at kør nedenstående celle, hvor der står print("Hello World!"). Dette skulle meget gerne skrive Hello World! under cellen.

In [4]:
print("Hello World!")

Hello World!


Vi kan også bruge print kommandoen til at printe tal, der bliver lagt sammen. I så fald bruger vi ikke anførelsestegn.

In [3]:
print(1+1)

2


Vi kan også trække tal fra hinanden, gange dem sammen, dividere dem, opløfte dem i en eksponenter og finde deres rest ved division

In [12]:
print(3 - 3)
print(3 * 3)
print(3 / 3)
print(3 ** 3)
print(3 % 3)

0
9
1.0
27
0


Hvis vi ville udskrive 1+1 ordret og ikke beregne noget, skulle der
anførselstegn omkring udtrykket, altså print("1+1").

In [4]:
print("1+1")

1+1


\n kan bruges for at lave linjeskift, og komma kan sættes mellem forskellige ting man vil printe.

In [10]:
print("første linje \n anden linje")
print("tredje linje", "\n", "fjerde linje")

første linje 
 anden linje
tredje linje 
 fjerde linje


Nogle nogle gange vil vi gerne forklare noget kode, dette gør vi
med såkaldte kommentarer. Så snart der er et # på en linje, ses resten
som en kommentar, som computeren ikke læser som kommandoer.

In [ ]:
# Dette er en kommentar
print("skibidi kommentar")
# print("yes yes")

### Opgaver:

##### Opgave 1.1
Skriv et program der printer "Hej verden!"

In [ ]:
# Din kode her

##### Opgave 1.2
Skriv et program der printer "Hej DIT NAVN", hvor du har
indsat dit eget navn.

In [ ]:
# Din kode her

##### Opgave 1.3
Skriv et program der printer resultatet af $2+2$ (uden at tallet
$4$ indgår i din kode).

In [ ]:
# Din kode her

# 2: Variabler

Fundamentet for at manipulere med data og lave udregninger ligger i at
kunne gemme resultater og værdier for så at kunne bruge dem igen senere.

Det at gemme værdier gøres med såkaldte variabler. Begrebet kender
læseren nok allerede fra matematik mere præcist aritmetik. Fra
Pythagoras sætning har vi eksempelvis $a^2 + b^2 = c^2$. Her er $a$, $b$ og $c$ variabler, der alle
sammen repræsenterer nogle (ukendte) værdier. Normalt haves eksempelvis $a$ og
$b$, og derudfra vil man så udregne $c^2$. Hvis eksempelvis at $a = 3$ og $b = 4$, så får
man $a^2 + b^2 = 25$. Det samme kan skrives i Python som:

In [2]:
a = 3
b = 4
c_i_anden = a ** 2 + b ** 2
print(c_i_anden)

25


I programmering skal “=” ikke forstås som et faktisk lighedstegn, som man kender fra matematik. Det
findes faktisk ikke i Python. Det skal derimod ses som en tildeling: Man
tildeler variablen, der står på venstre side af lighedstegnet, værdien af det,
der står på højre side.

Man kan godt tildele den samme variabel flere forskellige værdier, da der er et
lineært flow i udførelsen af programmet fra toppen af koden til bunden af
koden. Når først værdien af en variabel er blevet sat, så forbliver den ved
den tildelte værdi, indtil den bliver tildelt en ny værdi. Man kan eksempelvis skrive

In [1]:
a = 4
b = 5
c = a * a + b * b
print(c)
b = 1000
print(c)
c = a * a + b * b
print(c)

41
41
1000016


Man kunne måske forestille sig, at man ville have fået 41, 1000016,
1000016 som resultat, grundet at når værdien af b bliver ændret, så burde
c også opdateres. Men inden for programmering er det ikke tilfældet: c får
tildelt værdien af det på højre side af lighedstegnet og ikke selve
udtrykket. Derfor opdateres c ikke, når b opdateres

Som en kort notits kan man ikke bare kalde sine variabler hvad som helst.
Reglerne for navnene på en variabel i Python er at:

- Det skal starte med et bogstav eller en underscore.
- Derefter kan der være en vilkårlig sekvens af bogstaver, tal og under-
scores.
- Variabelnavne skelner mellem store og små bogstaver. F.eks. er Kage
ikke det samme som kage.

Variabler kan indeholde meget andet end tal. De kan faktisk indeholde alle
former for datatyper. Datatyper kan være mange forskellige ting, eksempelvis arrays
og booleske værdier, som der vil blive set på senere.

En af de mere fundamentale datatyper er “strenge” (eller "strings" på engelsk), hvilket er en
programmørs måde at arbejde med tekst på. Det kommer fra det engelske
“a string of characters”, på dansk, “en streng af tegn”. Vi kan eksempelvis skrive:

In [ ]:
variabel1 = "Hello"
variabel2 = "world!"
print(variabel1)
print(variabel2)

Hvis man indkapsler noget med citationstegn (enten "" eller '') i Python, så vil det altså blive registreret
som tekst. Aritmetiske operationer som addition kan også bruges på strenge.

Man kan eksempelvis skrive:

In [ ]:
variabel1 = "Hello"
variabel2 = "world!"
DenPerfekteVariabel = variabel1 + " " + variabel2
print(DenPerfekteVariabel)

In [ ]:
Hvis man laver komma mellem ting i sit print udtryk, sammensætter
Python det selv med mellemrum. Således:

In [4]:
variabel1 = "Hvad er svaret på livet, universet og alting?"
variabel2 = 42
print(variabel1, variabel2)

Hvad er svaret på livet, universet og alting? 42


Programmering involverer meget aritmetik. Derfor implementerer Python også de normale matematiske operatorer: plus, minus, gange og division. Disse ses brugt nedenfor:

| Operator | Beskrivelse                                                          |
| -------- | -------------------------------------------------------------------- |
| `+`      | Lægger to tal sammen (addition/sum).                                 |
| `-`      | Trækker højre operand fra venstre (subtraktion/differens).           |
| `*`      | Ganger to tal (multiplikation/produkt).                              |
| `/`      | Dividerer og returnerer en float (reel division/kvotient).           |
| `//`     | Heltalsdivision med afrunding nedad.                                 |
| `%`      | Returnerer resten af en division (modulus).                          |
| `**`     | Ophøjer venstre operand til højre operands potens.                   |


In [8]:
PlusDemonstration = 22 + 7
MinusDemonstration = 7 - 22
GangeDemonstration = 22 * 7
DivisionDemonstration = 22/7.0
PotensDemonstration = 22 ** 2
print(PlusDemonstration)
print(MinusDemonstration)
print(GangeDemonstration)
print(DivisionDemonstration)
print(PotensDemonstration)

29
-15
154
3.142857142857143
484


Bemærk her, at der er forskel på heltal (datatypen "integers") og vilkårlige relle tal (datatypen "floats"), hvorfor der i ovenstående eksempel bliver divideret med 7.0 i stedet for 7. Hvorfor det hele er skruet sådan sammen behøver vi ikke bekymre os om. Vi går groft sagt kun op i reelle tal, så bare skriv eksempelvis 2.0 i stedet for 2 i dine programmer, hvis du får fejlmeddelelser.

Der er normal præcedens for de forskellige operatorer. Det vil sige gange
og division binder stærkest efterfulgt af plus og minus. Putter man noget i
parenteser, så udregnes det altid først. Mellemrum mellem matematiske operationer ændrer ikke på andet end læselighed.

In [7]:
ParantesDemonstration = (2+2)/ (4-3    +11)

print(ParantesDemonstration)

0.3333333333333333


### Opgaver

##### Opgave 2.1
Ændr værdien af variablen x så der printes 3.

In [6]:
x = 1
print(x)

1


##### Opgave 2.2
Ændr værdien af variablen x så der printes 7.

In [5]:
x = 1
y = x + 2
print(y)

3


##### Opgave 2.3
Lav en variabel med navnet foo og værdien 1337. Derefter
print variablen. 

In [ ]:
# Din kode her

##### Opgave 2.4
Lav en variabel med navnet kage og værdien 12 derefter print den. Lav nu igen variablen kage, men denne gang giv den værdien "Joe Mama" og print den, sådan at hele programmet printer variablen kage 2 gange; først med værdien 12 derefter teksten Joe Mama. 

##### Opgave 2.5.
Følgende kode skulle printe "Hej med dig", men der er et problem. Find problemet og ret det.

In [ ]:
foo = "Hej med dig"
print("foo")

##### Opgave 2.6
Lav to variabler og giv dem forskellige tal-værdier. Lav så 5
print sætninger, hvor du bruger henholdsvis:
- Plus
- Minus
- Gange
- Division
- Potens

In [ ]:
# Din kode her

##### Opgave 2.7
Lav et program der beregner $c$ i Pythagoras sætning. Du skal som minimum
bruge to variabler til henholdsvis $a$ og $b$.
(Hint: du kan tage kvadratrod ved at opløfte med $\frac{1}{2}$)

In [ ]:
# Din kode her

# 3: Beslutninger

Man kan komme langt med variabler og regneoperationer i Python, men
programmet kan ikke udelukkende anvende ens udregninger til at foretage
beslutninger af sig selv. Man skal derimod instruere computeren i, hvilken
beslutning der skal tages på baggrund af den information og de beregninger,
der går forud i programmet. En beslutning kan eksempelvis være:

*Hvis butikken har sandwichkiks, så skal vi have tyve pakker af
dem, ellers skal vi brokke os højlydt ved kassen.*

Her har vi et boolsk udtryk: “hvis butikken har sandwichkiks”. Dette
udtryk kan enten være sandt eller falsk. Dernæst har vi, hvad vi gør, hvis
udtrykket er sandt: “så skal vi have tyve pakker af dem”. Og til sidst har
vi, hvad vi gør, hvis udtrykket er falsk: “ellers skal vi brokke os højlydt ved
kassen.”

Denne struktur bliver ofte set i programmering i form af if-else-strukturer.
I Python vil det forrige udtryk tage følgende form:

In [ ]:
har_sandwich_kiks = True

if har_sandwich_kiks:
    print("Køb tyve pakker.")
else:
    print("Brok dig højlydt ved kassen.")

Formen er altså:

In [ ]:
# if <boolsk udtryk>:
#     Hvis sandt
# else:
#     Hvis falsk

Bemærk if linjen og else linjen afsluttes med et kolon, og næste
linje er rykket ind i forhold til de andre. Den form følges også med andre
keywords i Python - et kolon er altid efterfulgt af øget indrykning (Engelsk:
indentation) på den følgende linje. Når noget er på samme indrykningsniveau, så hænger det sammen, dvs. enten kører det
hele, eller også bliver det hele sprunget over. Man kan også kalde det for en blok eller et scope.

Et boolsk udtryk kan tage mange former i Python. De simpleste er bare boolske værdier, som er True og False, hvilket henholdsvis betyder
sandt og falsk. Et boolsk udtryk kan opnås ved hjælp af sammenligninger af tal. Her er nogle eksempler:

In [5]:
12 == 12 # True - 12 er lig 12.
12 == 5 # False - 12 er lig 5.
5 < 10 # True - 5 er mindre end 10.
10 > 12 # False - 10 er større end 12.
10 <= 15 # True - 10 er mindre end eller lig med 15.
10 >= 10 # True - 10 er større end eller lig med 10.
5 != 5 # False - 5 er ikke lig 5.
10 != 5 # True - 10 er ikke lig 5.

True

Man kan foretage en negering af en boolsk værdi. Altså sandt bliver til
falsk og falsk bliver til sandt. Dette tager følgende form:

In [11]:
not True # False - Ikke sandt.
not False # True - Ikke falsk.

True

Man kan også sammenligne boolske udtryk med hinanden. Dette gøres
ved hjælp af de binære and - og or -konstruktioner. and sammenligner to
boolske værdier og returnerer sandt, hvis begge værdier er sande. or sam-
menligner to boolske værdier og returnerer sandt, hvis én af værdierne er
sand. Ved at kombinere disse kan man opbygge nogle mere intelligente udtryk.
Her kommer nogle eksempler:

In [14]:
# Du skal være mindst atten år eller være sammen med begge dine forældre.
alder = 17
med_mor = True
med_far = False
alder >= 18 or (med_mor and med_far)

False

In [13]:
# 2 skal gå op i x og x skal være større end 10 eller lig med 6.
# (’x % 2 == 0’ betyder at 2 går op i x)
x = 3
x % 2 == 0 and (x > 10 or x == 6)

False

In [17]:
# Computeren skal være til opladning eller have mindst 35% strøm og ikke være blevet brugt i over en time.
lader = 0.3
power = 3
idleMinutes = 300
(lader or power >= 0.35) and (idleMinutes > 60)

True

Man kan have det følgende mønster:

*Hvis butikken har Coca Cola, så køb det, ellers hvis de har Pepsi,
så køb det, ellers køb en billig Cola ingen gider at drikke lule.*

Dette kunne skrives med følgende struktur

In [ ]:
# if har_coca_cola:
#     Køb Coca Cola.
# else:
#     if har_pepsi:
#         Køb Pepsi.
# else:
#     Køb billig Cola.

Dog eksisterer en forkortende struktur i mange programmeringssprog, inklusiv i Python, som hedder else-if. Med denne struktur kan man skrive det
forrige således:

In [ ]:
# if har_coca_cola:
#     Køb Coca Cola.
# elif har_pepsi:
#     Køb Pepsi.
# else:
#     Køb billig Cola.

Her er et eksempel

In [8]:
yoloswag = 12

if yoloswag == 10:
    print("tralalero tralala")
elif yoloswag == 12:
    print("bombardino crocodilo")

bombardino crocodilo


### Opgaver

##### Opgave 3.1
Skriv et program hvor der printes "STOR", hvis en variabel er
større end 100.

In [ ]:
# Din kode her

##### Opgave 3.2
Følgende kode skulle printe "Yay", når x er nøjagtigt lig 4,
og "Done" til sidst uanset hvad, men der er en fejl. Find fejlen og ret den.

In [18]:
x = 4
if x = 4:
    print("Yay")
print("Done")

SyntaxError: invalid syntax. Maybe you meant '==' or ':=' instead of '='? (2896845246.py, line 2)

Opgave 3.3
Følgende kode skulle printe "Yay", hvis x er skarpt mindre
end 5, og under alle omstændigheder printe "Done", men der er en fejl. Find fejlen og ret den

In [ ]:
x = 2
if x < 5:
print("Yay")
print("Done")

##### Opgave 3.3
Følgende kode skulle printe "A"hvis x er større end 5 og ellers
"B", men der er en fejl. Find og ret fejlen

In [ ]:
x = 8
if x > 5:
    print("A")
    else:
    print("B")

##### Opgave 3.4
Lav et program med variablerne alder og med_voksen, som
printer "Velkommen til forlystelsen", hvis man er over 10 år, eller hvis man
er med en voksen.

In [ ]:
# Din kode her

##### Opgave 3.5
Lav et program med variablerne a og b, der printer strengen "Det største tal er:" efterfulgt variablen med størst værdi på samme linje

In [ ]:
# Din kode her

##### Opgave 3.6
Lav et program med variablerne gangsta og rap, der printer strengen "crazy mf named ice cube", hvis begge variable er større end 5.

In [ ]:
# Din kode her

##### Opgave 3.7
Lav et program med variablerne rip rap og rup, der printer "onkel anders", hvis en af de tre variable er mindre end eller lig med 74. Hvis alle tre variable er forskellige fra tallet -2 skal programmet i stedet printe "onkel joakim".

In [ ]:
# Din kode her

##### Opgave 3.8
En andengradsligning er en ligning på formen $ax^2 +bx+c =
0$, hvor målet er at finde x. Skriv et program, der definerer 3 variable $a$, $b$ og $c$ og udskriver
løsningerne til ligningen, hvis de findes.

Antallet af løsninger kan bestemmes ud fra diskriminanten $d = b^2 − 4ac$
som følger:

- $d < 0$: Der findes ingen reelle løsninger.
- $d = 0$: Der findes én løsning $x = \frac{-b}{2a}−b$
- $d > 0$: Der findes to løsninger givet ved $x = \frac{−b±\sqrt{d}}{2a}$

In [ ]:
# Din kode her